src/main/java/training/chessington/model/Board.java

```diff
@@ -53,27 +54,44 @@ public void placePiece(Coordinates coords, Piece piece) {
     }
 
     public boolean moveIsBlocked(Move move) {
-        
-        int fromRow = move.getFrom().getRow();
-        int fromCol = move.getFrom().getCol();
-        int toRow = move.getTo().getRow();
-        int toCol = move.getTo().getCol();
-        
-        boolean isHorizontal = fromRow == toRow;
-        boolean isVertical = fromCol == toCol;
-        boolean isDiagonal = (fromRow - toRow) == (fromCol - toCol);
-        
-        if (isHorizontal) {
-            return linearMoveIsBlocked(fromRow, fromCol, toCol, false);
-        } else if (isVertical) {
-            return linearMoveIsBlocked(fromCol, fromRow, toRow, true);
+        Function<Coordinates, Coordinates> directionalStep = this.calculateDirectionalStep(move.getFrom(), move.getTo());
+        return this.linearMoveIsBlocked(move, directionalStep);
+    }
+
+    private Function<Coordinates, Coordinates> calculateDirectionalStep(Coordinates start, Coordinates end) {
+        if (start.equals(end)) {
+            return (position) -> position;
+        } else if (start.getRow() == end.getRow()) {
+            int colDisplacement = end.getCol() - start.getCol();
+            int direction = colDisplacement / Math.abs(colDisplacement);
+            return (position) -> position.plus(0, direction);
+        } else if (start.getCol() == end.getCol()) {
+            int rowDisplacement =  end.getRow() - start.getRow();
+            int direction = rowDisplacement / Math.abs(rowDisplacement);
+            return (position) -> position.plus(direction, 0);
+        } else {
+            int rowDisplacement = end.getRow() - start.getRow();
+            int colDisplacement = end.getCol() - start.getCol();
+            int rowDirection = rowDisplacement / Math.abs(rowDisplacement);
+            int colDirection = colDisplacement / Math.abs(colDisplacement);
+            return (position) -> position.plus(rowDirection, colDirection);
+        }
+    }
+
+    private boolean linearMoveIsBlocked(Move move, Function<Coordinates, Coordinates> directionalStep) {

```

Comment:
This method seems like it takes too many parameters - can the step not be worked out from the move? As it stands, you could specify a directional step that doesn't actually match the move (would be an easy mistake to make!) and then it wouldn't terminate, I think...

Reply:
Yeah, I'm trying to follow the "each function should step down one level of abstraction" rule, but I'm not sure how to do that without putting that while loop logic in a separate method, which requires access to the start and end points and the directional step, unless it calculates it itself (which breaks the abstraction level rule).